File: 02-linkedlists/ts/singly-linked-list/main.js
"use strict";
exports.__esModule = true;
var ListNode = /** @class */ (function () {
    function ListNode(data, next) {
        this.data = data;
        this.next = next;
    }
    return ListNode;
}());
var SinglyLinkedList = /** @class */ (function () {
    function SinglyLinkedList() {
        this.dhead = new ListNode(null, null);
        this.size = 0;
    }
    SinglyLinkedList.prototype.getSize = function () {
        return this.size;
    };
    SinglyLinkedList.prototype.isEmpty = function () {
        return this.size == 0;
    };
    SinglyLinkedList.prototype.front = function () {
        var _a;
        if (this.isEmpty()) {
            return null;
        }
        return (_a = this.dhead.next) === null || _a === void 0 ? void 0 : _a.data;
    };
    SinglyLinkedList.prototype.back = function () {
        if (this.isEmpty()) {
            return null;
        }
        var curr = this.dhead;
        while (curr.next != null) {
            curr = curr.next;
        }
        return curr.data;
    };
    /* pos: 1-based */
    SinglyLinkedList.prototype.value_at = function (pos) {
        if (this.isEmpty()) {
            console.log("List is empty..");
            return null;
        }
        if (pos <= 0 || pos > this.size) {
            console.log("Invalid value... position:", pos);
            return null;
        }
        var curr = this.dhead;
        while (pos && curr.next) {
            curr = curr.next;
            --pos;
        }
        return curr.data;
    };
    SinglyLinkedList.prototype.push_front = function (data) {
        var newnode = new ListNode(data, this.dhead.next);
        this.dhead.next = newnode;
        this.size++;
    };
    SinglyLinkedList.prototype.push_back = function (data) {
        if (this.isEmpty()) {
            this.push_front(data);
            return null;
        }
        var curr = this.dhead;
        while (curr.next != null) {
            curr = curr.next;
        }
        var newnode = new ListNode(data, null);
        curr.next = newnode;
        this.size++;
    };
    SinglyLinkedList.prototype.pop_front = function () {
        if (this.isEmpty()) {
            console.log("List is empty...");
            return null;
        }
        var curr = this.dhead.next;
        if (this.size == 1) {
            this.dhead.next = null;
            this.size--;
            return curr === null || curr === void 0 ? void 0 : curr.data;
        }
        else {
            var temp = this.dhead.next;
            this.dhead.next = temp === null || temp === void 0 ? void 0 : temp.next;
            var data = temp === null || temp === void 0 ? void 0 : temp.data;
            temp = null;
            this.size--;
            return data;
        }
    };
    SinglyLinkedList.prototype.pop_back = function () {
        var _a, _b;
        if (this.isEmpty()) {
            console.log("List is empty...");
            return null;
        }
        if (this.size == 1) {
            return this.pop_front();
        }
        var curr = this.dhead;
        while ((_a = curr.next) === null || _a === void 0 ? void 0 : _a.next) {
            curr = curr.next;
        }
        var data = (_b = curr.next) === null || _b === void 0 ? void 0 : _b.data;
        curr.next = null;
        this.size--;
        return data;
    };
    /* pos: 1-based */
    SinglyLinkedList.prototype.insert = function (data, pos) {
        if (pos <= 0 || pos > this.size + 1) {
            console.log("Invalid value... position:", pos);
            return null;
        }
        if (pos == 1) {
            this.push_front(data);
        }
        else if (pos == this.size + 1) {
            this.push_back(data);
        }
        else {
            var curr = this.dhead;
            --pos;
            while (pos && curr.next) {
                curr = curr.next;
                --pos;
            }
            var newnode = new ListNode(data, curr.next);
            curr.next = newnode;
            this.size++;
        }
    };
    /* pos = 1-based */
    SinglyLinkedList.prototype.erase = function (pos) {
        var _a, _b;
        if (pos <= 0 || pos > this.size) {
            console.log("Invalid value... position:", pos);
            return null;
        }
        if (pos == 1) {
            return this.pop_front();
        }
        if (pos == this.size) {
            return this.pop_back();
        }
        --pos;
        var curr = this.dhead;
        while (pos && curr.next) {
            curr = curr.next;
            --pos;
        }
        var data = (_a = curr.next) === null || _a === void 0 ? void 0 : _a.data;
        curr.next = (_b = curr.next) === null || _b === void 0 ? void 0 : _b.next;
        this.size--;
        return data;
    };
    SinglyLinkedList.prototype.remove_value = function (data) {
        if (this.isEmpty()) {
            console.log("List is empty...");
            return false;
        }
        var curr = this.dhead;
        while (curr.next) {
            if (curr.next.data == data) {
                var temp = curr.next;
                curr.next = curr.next.next;
                temp.next = null;
                this.size--;
                return true;
            }
            curr = curr.next;
        }
        return false;
    };
    SinglyLinkedList.prototype.reverse = function () {
        var prev = null;
        var curr = this.dhead.next;
        while (curr != null) {
            var next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        this.dhead.next = prev;
    };
    SinglyLinkedList.prototype.print = function () {
        if (this.isEmpty()) {
            return null;
        }
        var curr = this.dhead.next;
        var output = '';
        var front_back = "f:".concat(this.front(), ", b:").concat(this.back(), ")");
        while (curr) {
            output += "".concat(curr.data, " ");
            curr = curr.next;
        }
        output += "(size:".concat(this.getSize(), ", ").concat(front_back);
        console.log(output);
    };
    return SinglyLinkedList;
}());
var prompt = require("prompt-sync")({ sigint: true });
var list = new SinglyLinkedList();
console.log("------- Operations ------");
console.log(" 1. push_back      2. push_front");
console.log(" 3. pop_back       4. pop_front");
console.log(" 5. insert");
console.log(" 6. erase          7. remove_value");
console.log(" 8. front          9. back");
console.log("10. size          11. isEmpty");
console.log("12. print         13. reverse");
console.log("14. valueAt        0. Exit");
var isRunning = true;
while (isRunning) {
    var answer = prompt("> ");
    switch (answer) {
        case "1": {
            var data = prompt("push back: ");
            list.push_back(parseInt(data));
            list.print();
            break;
        }
        case "2": {
            var data = prompt("push front: ");
            list.push_front(parseInt(data));
            list.print();
            break;
        }
        case "3": {
            list.pop_back();
            list.print();
            break;
        }
        case "4": {
            list.pop_front();
            list.print();
            break;
        }
        case "5": {
            var data = prompt("Insert data: ");
            var pos = prompt("at position (1-based): ");
            list.insert(data, pos);
            list.print();
            break;
        }
        case "6": {
            var pos = prompt("Erase a value at position (1-based): ");
            list.erase(pos);
            list.print();
            break;
        }
        case "7": {
            var data = prompt("Remove a value: ");
            list.remove_value(data);
            list.print();
            break;
        }
        case "8": {
            console.log("Front:", list.front());
            break;
        }
        case "9": {
            console.log("Back:", list.back());
            break;
        }
        case "10": {
            console.log("Size:", list.getSize());
            break;
        }
        case "11": {
            if (list.isEmpty()) {
                console.log("List is empty.");
            }
            else {
                console.log("List is not empty.");
            }
            break;
        }
        case "12": {
            list.print();
            break;
        }
        case "13": {
            list.reverse();
            list.print();
            break;
        }
        case "14": {
            var pos = prompt("Value at position (1-based): ");
            console.log(list.value_at(pos));
            break;
        }
        case "0": {
            console.log("Bye!");
            isRunning = false;
        }
    }
}

File: 02-linkedlists/singly-linked-list/main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define TRUE 1
#define FALSE 0 

#include "debug.h"
#include "sll.h"

int main(void) {
    unsigned seed = time(0);
    srand(seed);

    SLL *list = init();

    #ifdef TEST_SIZE
        printf("size: %d\n", list->size);
        printf("\n");
    #endif

    #ifdef TEST_EMPTY
        if(empty(list)) {
            printf("List is empty\n");
        } else {
            printf("List is not empty\n");
        }
        printf("\n");
    #endif

    #ifdef TEST_PUSH_FRONT
        for(int i=0; i<10; ++i) {
            push_front(list, i);
            print(list);
        }
        printf("\n");
    #endif

    #ifdef TEST_PUSH_BACK
        for(int i=0; i<10; ++i) {
            push_back(list, i);
            print(list);
        }
        printf("\n");
    #endif

    #ifdef TEST_POP_FRONT
        for(int i=0; i<5; ++i) {
            pop_front(list);
        }
        printf("\n");
    #endif 

    #ifdef TEST_ERASE
        for(int i=0; i<5; ++i) {
            erase(list, rand()%10);
            print(list);
        }
        printf("\n");
    #endif

    #ifdef TEST_INSERT
        for(int i=0; i<10; ++i) {
            insert(list, rand()%10, rand() % 100);
            print(list);
        }
        printf("\n");
    #endif

    #ifdef TEST_POP_BACK
        for(int i=0; i<5; ++i) {
            pop_back(list);
        }
        printf("\n");
    #endif

    #ifdef TEST_FRONT
        printf("front: %d\n", front(list));
        printf("\n");
    #endif

    #ifdef TEST_BACK
        printf("back: %d\n", back(list));
        printf("\n");
    #endif

    #ifdef TEST_VALUE_AT
        printf("value_at(0): %d\n", value_at(list, 0));
        printf("\n");
    #endif

    #ifdef TEST_PRINT_FROM_N
        for(int i=0; i<list->size+1; ++i) {
            print_from_n(list, i);
        }
        printf("\n");
    #endif

    #ifdef TEST_REMOVE_VALUE
        remove_value(list, 10);
        print(list);
        remove_value(list, 2);
        print(list);
        remove_value(list, 18);
        print(list);
        remove_value(list, 13);
        print(list);
        printf("\n");
    #endif

    #ifdef TEST_REVERSE
        for(int i=0; i<10; ++i) {
            push_back(list, rand() % 100);
        }
        printf("\n Regular: ");
        print(list);
        reverse(list);
        printf("\n Reversed: ");
        print(list);
    #endif 

    _free(&list);

    return 0;
}

SLL* init() {
    SLL *list = (SLL *)malloc(sizeof(SLL));

    list->sentinel = new_node(-1, NULL);
    list->size = 0;

    return list;
}

void _free(SLL **list) {
#ifdef DEBUG
    printf("[[ inside '_free()' ]]\n");
#endif
    while(!empty(*list)) {
        pop_front(*list);
    }

    free((*list)->sentinel);
    free(*list);
    *list = NULL;
}

void print(SLL *list) {
#ifdef DEBUG
    printf("[[ inside 'print()' ]]\n");
#endif
    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return;
    }

    ListNode *temp = list->sentinel;
    while(temp->next) {
        temp = temp->next;
        printf("%d ", temp->data);
    }
    printf("\n");
}

void print_from_n(SLL *list, int index) {
#ifdef DEBUG
    printf("[[ inside 'print_from_n(%d)' ]]\n", index);
#endif
    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return;
    }
    
    if(index < 0 || index >= list->size) {
        printf("ERROR: index out of range\n");
        return;
    }

    ListNode *temp = list->sentinel->next;

    for(int i=0; i<index; ++i) {
        temp = temp->next;
    }

    while(temp) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    
    printf("\n");
}

void reverse(SLL *list) {
#ifdef DEBUG
    printf("[[ inside reverse ]]\n");
#endif
    if(list->size == 0) {
        printf("ERROR: empty list\n");
        return;
    }

    ListNode *prev = NULL;
    ListNode *curr = list->sentinel->next;
    
    while(curr != NULL){
        ListNode *next = curr->next;
        curr->next=prev;
        prev = curr;
        curr=next;
    }
    list->sentinel->next = prev;
}

int size(SLL *list) {
    return list->size;
}

_Bool empty(SLL *list) {
    return list->size == 0;
}

void push_front(SLL *list, int data) {
#ifdef DEBUG
    printf("[[ inside 'push_front(%d)' ]]\n", data);
#endif
    ListNode *newNode = new_node(data, list->sentinel->next);
    list->sentinel->next = newNode;
    ++(list->size);
}

int pop_front(SLL *list) {
#ifdef DEBUG
    printf("[[ inside 'pop_front' ]]\n");
#endif
    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return list->sentinel->data;
    }

    ListNode *temp = list->sentinel->next;
    int data = temp->data;
    list->sentinel->next = temp->next;
    free(temp);

    --(list->size);

#ifdef DEBUG
    printf("[[ popping %d ]]\n", data);
#endif
    return data;
}

void push_back(SLL *list, int data) {
#ifdef DEBUG
    printf("[[ inside 'push_back(%d)' ]]\n", data);
#endif
    ListNode *curr = list->sentinel;

    while(curr->next) {
        curr = curr->next;
    }

    curr->next = new_node(data, NULL);
    ++(list->size);
}

int pop_back(SLL *list) {
#ifdef DEBUG
    printf("[[ inside 'pop_back' ]]\n");
#endif

    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return list->sentinel->data;
    }

    ListNode *curr = list->sentinel;
    while(curr->next->next) {
        curr = curr->next;
    }

    ListNode *t = curr->next;
    int data = t->data;
    curr->next = t->next;
    free(t);

#ifdef DEBUG
    printf("[[ popping: %d ]]\n", data);
#endif

    --(list->size);
    return data;
}

void insert(SLL *list, int index, int data) {
#ifdef DEBUG
    printf("[[ inside 'insert(index:%d, data:%d)' ]]\n", index, data);
#endif

    if(index < 0 || index > list->size) {
        printf("ERROR: index out of range.\n");
        return;
    }

    if(index == 0) {
        push_front(list, data);
    } else if(index == list->size) {
        push_back(list, data);
    } else {
        ListNode *curr = list->sentinel;
        for(int i=0; i<index; ++i) {
            curr = curr->next;
        }

        ListNode *newNode = new_node(data, curr->next);
        curr->next = newNode;

        ++(list->size);
    }
}

void erase(SLL *list ,int index) {
#ifdef DEBUG
    printf("[[ inside 'erase(index:%d)' ]]\n", index);
#endif

    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return;
    }
    if(index < 0 || index > list->size) {
        printf("ERROR: index out of range.\n");
        return;
    }

    if(index == 0) {
        pop_front(list);
    } else if(index == list->size-1) {
        pop_back(list);
    } else {
        ListNode *curr = list->sentinel;

        for(int i=0; i<index; ++i) {
            curr = curr->next;
        }

        ListNode *t = curr->next;
        curr->next = t->next;
        free(t);
        --(list->size);
    }
}

_Bool remove_value(SLL *list, int data) {
#ifdef DEBUG
    printf("[[ inside 'remove_value(data: %d)' ]]\n", data);
#endif

    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return FALSE;
    }

    ListNode *curr = list->sentinel;

    while(curr->next) {
        if(curr->next->data == data) {
        #ifdef DEBUG
            printf("[[ value found... deleting the node.. ]]\n");
        #endif
            ListNode *t = curr->next;
            curr->next = t->next;
            free(t);
            --(list->size);
            return TRUE;
        }

        curr = curr->next;
    }
    #ifdef DEBUG
        printf("[[ value NOT found... ]]\n");
    #endif
    return FALSE;
}

/* returns the value of the nth item (starting at 0 for first) */
/* 0-based indexing */
int value_at(SLL *list, int index) {
#ifdef DEBUG
    printf("[[ inside 'value_at(list, %d)' ]]\n", index);
#endif
    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return list->sentinel->data;
    }
    if(index < 0 || index >= list->size) {
        printf("ERROR: index out of range.\n");
        return list->sentinel->data;
    }
    ListNode *curr = list->sentinel->next;

    for(int i=0; i<index; ++i) {
        curr = curr->next;
    }

    return curr->data;
}

int front(SLL *list) {
#ifdef DEBUG
    printf("[[ inside 'front' ]]\n");
#endif
    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return list->sentinel->data;
    }

    return list->sentinel->next->data;
}

int back(SLL *list) {
#ifdef DEBUG
    printf("[[ inside 'back' ]]\n");
#endif
    if(list->size == 0) {
        printf("ERROR: list is empty.\n");
        return list->sentinel->data;
    }

    ListNode *temp = list->sentinel->next;

    while(temp->next) {
        temp = temp->next;
    }

    return temp->data;
}